### PR TITLE
[DeviceMSAN] Support MemorySanitizer for device offloading

### PR DESCRIPTION
**Pull request by AllanZyne**
LLVM: https://github.com/intel/llvm/pull/15955

### CODE REVIEW

File: source/loader/layers/sanitizer/asan/asan_ddi.cpp [hunk 1]
```diff
@@ -1888,6 +1870,7 @@ ur_result_t context_t::init(ur_dditable_t *dditable,
         result = ur_sanitizer_layer::asan::urGetUSMProcAddrTable(
             UR_API_VERSION_CURRENT, &dditable->USM);
     }
+    getContext()->logger.info("initMsanDDITable: {}", result);
```
Comment (zhaomaosu) [inline source/loader/layers/sanitizer/asan/asan_ddi.cpp:1873]: Should it be Asan? BTW 'initMsanDDITable' is not user friendly.

File: source/loader/layers/sanitizer/msan/msan_buffer.cpp [hunk 2]
```diff
@@ -0,0 +1,204 @@
+/*
+ *
+ * Copyright (C) 2024 Intel Corporation
+ *
+ * Part of the Unified-Runtime Project, under the Apache License v2.0 with LLVM Exceptions.
+ * See LICENSE.TXT
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ *
+ * @file asan_buffer.cpp
```
Comment (zhaomaosu) [inline source/loader/layers/sanitizer/msan/msan_buffer.cpp:9]:
```suggestion
 * @file msan_buffer.cpp
```

File: source/loader/layers/sanitizer/msan/msan_ddi.cpp [hunk 3]
```diff
@@ -0,0 +1,1524 @@
...
+    ur_adapter_handle_t *
+        phAdapters, ///< [out][optional][range(0, NumEntries)] array of handle of adapters.
+    ///< If NumEntries is less than the number of adapters available, then
+    ///< ::urAdapterGet shall only retrieve that number of platforms.
+    uint32_t *
+        pNumAdapters ///< [out][optional] returns the total number of adapters available.
+) {
+    auto pfnAdapterGet = getContext()->urDdiTable.Global.pfnAdapterGet;
+
+    setenv("NEOReadDebugKeys", "1", 1);
```
Comment (zhaomaosu) [inline source/loader/layers/sanitizer/msan/msan_ddi.cpp:78]: Added a fixme for these code?

File: source/loader/layers/sanitizer/asan/asan_ddi.cpp (hunk 1 above)
Comment (yingcong-wu) [inline source/loader/layers/sanitizer/asan/asan_ddi.cpp:1873]: Why it logs "Msan" in asan_ddi.cpp?

File: source/loader/layers/sanitizer/msan/msan_buffer.cpp (hunk 2 above)
Comment (yingcong-wu) [inline source/loader/layers/sanitizer/msan/msan_buffer.cpp:9]:
```suggestion
 * @file msan_buffer.cpp
```

File: source/loader/layers/sanitizer/msan/msan_interceptor.cpp [hunk 4]
```diff
@@ -0,0 +1,502 @@
+//===----------------------------------------------------------------------===//
+/*
+ *
+ * Copyright (C) 2023 Intel Corporation
```
Comment (zhaomaosu) [inline source/loader/layers/sanitizer/msan/msan_interceptor.cpp:4]:
```suggestion
 * Copyright (C) 2024 Intel Corporation
```

File: source/loader/layers/sanitizer/msan/msan_interceptor.cpp [hunk 5]
```diff
@@ -0,0 +1,502 @@
...
+    m_ContextMap.clear();
+
+    for (auto Adapter : m_Adapters) {
+        getContext()->urDdiTable.Global.pfnAdapterRelease(Adapter);
+    }
+}
+
+/// The memory chunk allocated from the underlying allocator looks like this:
+/// L L L L L L U U U U U U R R
+///   L -- left redzone words (0 or more bytes)
```
Comment (zhaomaosu) [inline source/loader/layers/sanitizer/msan/msan_interceptor.cpp:47]: memory sanitizer doesn't use redone, looks like the comments can be removed

File: source/loader/layers/sanitizer/msan/msan_interceptor.cpp [hunk 6]
```diff
@@ -0,0 +1,502 @@
...
+        UR_CALL(enqueueAllocInfo(DeviceInfo, Queue, AI));
+    }
+    AllocInfos.List.clear();
+
+    return UR_RESULT_SUCCESS;
+}
+
+ur_result_t MsanInterceptor::registerProgram(ur_program_handle_t Program) {
+    ur_result_t Result = UR_RESULT_SUCCESS;
+    do {
```
Comment (zhaomaosu) [inline source/loader/layers/sanitizer/msan/msan_interceptor.cpp:159]: remove loop

File: source/loader/layers/sanitizer/msan/msan_interceptor.cpp [hunk 7]
```diff
@@ -0,0 +1,502 @@
...
+    return UR_RESULT_SUCCESS;
+}
+
+bool ProgramInfo::isKernelInstrumented(ur_kernel_handle_t Kernel) const {
+    const auto Name = GetKernelName(Kernel);
+    return InstrumentedKernels.find(Name) != InstrumentedKernels.end();
+}
+
+ContextInfo::~ContextInfo() {
+    // Stats.Print(Handle);
```
Comment (zhaomaosu) [inline source/loader/layers/sanitizer/msan/msan_interceptor.cpp:453]: remove unused code?

File: source/loader/layers/sanitizer/msan/msan_ddi.cpp [hunk 8]
```diff
@@ -0,0 +1,1524 @@
...
+    setenv("AllocateHostAllocationsInHeapExtendedHost", "0", 1);
```
Comment (yingcong-wu) [inline source/loader/layers/sanitizer/msan/msan_ddi.cpp:79]: Could we add some comments here to help other understand why we need these keys?

File: source/loader/layers/sanitizer/msan/msan_interceptor.hpp [hunk 9]
```diff
@@ -0,0 +1,334 @@
+/*
+ *
+ * Copyright (C) 2023 Intel Corporation
```
Comment (zhaomaosu) [inline source/loader/layers/sanitizer/msan/msan_interceptor.hpp:3]:
```suggestion
 * Copyright (C) 2024 Intel Corporation
```

File: source/loader/layers/sanitizer/msan/msan_interceptor.hpp [hunk 10]
```diff
@@ -0,0 +1,334 @@
...
+};
+
+struct KernelInfo {
+    ur_kernel_handle_t Handle;
+    std::atomic<int32_t> RefCount = 1;
+
+    // lock this mutex if following fields are accessed
+    ur_shared_mutex Mutex;
+    std::unordered_map<uint32_t, std::shared_ptr<MemBuffer>> BufferArgs;
+    // std::unordered_map<uint32_t, std::pair<const void *, StackTrace>>
```
Comment (zhaomaosu) [inline source/loader/layers/sanitizer/msan/msan_interceptor.hpp:82]: remove unused code

File: source/loader/layers/sanitizer/msan/msan_interceptor.hpp [hunk 11]
```diff
@@ -0,0 +1,334 @@
...
+            this->LocalWorkSize =
+                std::vector<size_t>(LocalWorkSize, LocalWorkSize + WorkDim);
+        }
+    }
+    ~USMLaunchInfo();
+
+    ur_result_t initialize();
+};
+
+struct DeviceGlobalInfo {
```
Comment (zhaomaosu) [inline source/loader/layers/sanitizer/msan/msan_interceptor.hpp:173]: looks like this struct can be removed

File: source/loader/layers/sanitizer/sanitizer_common/sanitizer_shadow.hpp [hunk 12]
```diff
@@ -0,0 +1,19 @@
+/*
+ *
+ * Copyright (C) 2024 Intel Corporation
+ *
+ * Part of the Unified-Runtime Project, under the Apache License v2.0 with LLVM Exceptions.
+ * See LICENSE.TXT
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ *
+ * @file sanitizer_stacktrace.hpp
```
Comment (zhaomaosu) [inline source/loader/layers/sanitizer/sanitizer_common/sanitizer_shadow.hpp:9]: Why we need this header file?

File: source/loader/layers/sanitizer/msan/msan_interceptor.hpp (hunk 9 above)
Comment (yingcong-wu) [inline source/loader/layers/sanitizer/msan/msan_interceptor.hpp:3]:
```suggestion
 * Copyright (C) 2024 Intel Corporation
```

File: source/loader/layers/sanitizer/msan/msan_report.cpp [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+                                  ur_kernel_handle_t Kernel) {
+    const char *File = Report.File[0] ? Report.File : "<unknown file>";
+    const char *Func = Report.Func[0] ? Report.Func : "<unknown func>";
+    auto KernelName = GetKernelName(Kernel);
+
+    // Try to demangle the kernel name
+    KernelName = DemangleName(KernelName);
+
+    getContext()->logger.always(
+        "====WARNING: MemorySanitizer: use-of-uninitialized-value");
```
Comment (yingcong-wu) [inline source/loader/layers/sanitizer/msan/msan_report.cpp:33]: If we use Memory sanitizer here, should we update the DevASAN? To speak for myself, I think we can use "DeviceSanitizer" for all sanitizers.

File: source/loader/layers/sanitizer/msan/msan_shadow.hpp [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+    ur_context_handle_t Context{};
+
+    ur_device_handle_t Device{};
+
+    uptr ShadowBegin = 0;
+
+    uptr ShadowEnd = 0;
+};
+
+// Shadow Memory layout of CPU PVC device
```
Comment (yingcong-wu) [inline source/loader/layers/sanitizer/msan/msan_shadow.hpp:53]: Do you mean "CPU/PVC" or CPU Device?

File: source/loader/layers/sanitizer/msan/msan_shadow.hpp [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+struct MsanShadowMemoryPVC final : public MsanShadowMemoryGPU {
+    MsanShadowMemoryPVC(ur_context_handle_t Context, ur_device_handle_t Device)
+        : MsanShadowMemoryGPU(Context, Device) {}
+
+    uptr MemToShadow(uptr Ptr) override;
+
+    size_t GetShadowSize() override { return 0x8000'0000'0000ULL; }
+};
+
+/// Shadow Memory layout of GPU PVC device
```
Comment (yingcong-wu) [inline source/loader/layers/sanitizer/msan/msan_shadow.hpp:138]: Not matching the code.

File: source/loader/layers/sanitizer/sanitizer_common/sanitizer_shadow.hpp (hunk 12 above)
Comment (yingcong-wu) [inline source/loader/layers/sanitizer/sanitizer_common/sanitizer_shadow.hpp:9]:
```suggestion
 * @file sanitizer_shadow.hpp
```

**Review by yingcong-wu — APPROVED** [review]

**Review by zhaomaosu — APPROVED** [review]
lgtm

**Comment by AllanZyne** [comment]
Hi @oneapi-src/unified-runtime-maintain, internal review has done. Please review, thank you.

**Comment by AllanZyne** [comment]
CI "Build and test / E2E OpenCL / Start e2e job / Build SYCL, UR, run E2E" failed is due to opencl driver haven't upgraded to 2025.0.

**Comment by cdai2** [comment]
ping @oneapi-src/unified-runtime-maintain

**Comment by kbenzie** [comment]
> ping @oneapi-src/unified-runtime-maintain

https://github.com/intel/llvm/pull/15955 is *Waiting on code owner review from intel/dpcpp-clang-driver-reviewers, intel/dpcpp-esimd-reviewers, intel/dpcpp-tools-reviewers, intel/llvm-reviewers-runtime* until those code owners approve that PR this won't be merged.

**Comment by AllanZyne** [comment]
Hi @oneapi-src/unified-runtime-maintain, I have got almost all reviewers approve, could you please review this PR as well?
Thank you!

File: source/loader/layers/sanitizer/ur_sanddi.cpp [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+#include "asan/asan_ddi.hpp"
+#include "msan/msan_ddi.hpp"
+#include "ur_sanitizer_layer.hpp"
+
+namespace ur_sanitizer_layer {
+
+ur_result_t context_t::init(ur_dditable_t *dditable,
+                            const std::set<std::string> &enabledLayerNames,
+                            [[maybe_unused]] codeloc_data codelocData) {
+    if (enabledLayerNames.count("UR_LAYER_ASAN")) {
```
Comment (pbalcer) [inline source/loader/layers/sanitizer/ur_sanddi.cpp:22]:
```
bool asan_enabled = enabledLayerNames.count("UR_LAYER_MSAN");
bool msan_enabled = enabledLayerNames.count("UR_LAYER_MSAN");

if (asan_enabled && msan_enabled) {
  getContext()->logger.warning("Enabling MSAN and ASAN at the same time is not supported.")
  return UR_RESULT_SUCCESS;
} else if (asan_enabled) {
  ...
}
```

Just a suggestion

File: source/loader/layers/sanitizer/msan/msan_interceptor.cpp [hunk 17]
```diff
@@ -0,0 +1,493 @@
...
+    }
+
+    return Result;
+}
+
+ur_result_t MsanInterceptor::unregisterProgram(ur_program_handle_t) {
+    return UR_RESULT_SUCCESS;
+}
+
+ur_result_t MsanInterceptor::registerSpirKernels(ur_program_handle_t Program) {
```
Comment (pbalcer) [inline source/loader/layers/sanitizer/msan/msan_interceptor.cpp:166]: This, and a few other functions in this file, are nearly straight copy-paste from asan interceptor.

I know there's time pressure now, but please create a high-priority ticket to refactor this post merge. Maybe there should be a common base class for sanitizers that contains the shared code?

File: source/loader/layers/sanitizer/msan/msan_interceptor.cpp [hunk 18]
```diff
@@ -0,0 +1,493 @@
...
+    }
+    return nullptr;
+}
+
+ur_result_t MsanInterceptor::prepareLaunch(
+    std::shared_ptr<DeviceInfo> &DeviceInfo, ur_queue_handle_t Queue,
+    ur_kernel_handle_t Kernel, USMLaunchInfo &LaunchInfo) {
+    auto Program = GetProgram(Kernel);
+
+    do {
```
Comment (pbalcer) [inline source/loader/layers/sanitizer/msan/msan_interceptor.cpp:351]: please remove this loop.
I also noticed that the same thing is done in the same asan function. Please remove that as well.

File: source/loader/layers/sanitizer/msan/msan_options.cpp [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+#include "ur_sanitizer_layer.hpp"
+
+#include <algorithm>
+#include <cstring>
+#include <stdexcept>
+
+namespace ur_sanitizer_layer {
+namespace msan {
+
+MsanOptions::MsanOptions() {
```
Comment (pbalcer) [inline source/loader/layers/sanitizer/msan/msan_options.cpp:25]: This is a near 1:1 copy/paste from asan_options.

File: source/loader/layers/sanitizer/msan/msan_shadow.cpp [hunk 20]
```diff
@@ -0,0 +1,289 @@
...
+            return UR_RESULT_ERROR_UNKNOWN;
+        }
+        ShadowBegin = ShadowEnd = 0;
+        return UR_RESULT_SUCCESS;
+    }();
+    return Result;
+}
+
+uptr MsanShadowMemoryCPU::MemToShadow(uptr Ptr) {
+    return Ptr ^ 0x500000000000ULL;
```
Comment (pbalcer) [inline source/loader/layers/sanitizer/msan/msan_shadow.cpp:109]: create a define for the magic number

**Review by pbalcer — COMMENTED** [review]
Mostly lgtm. My biggest concern is the large amounts of copy/pasting that the new msan is doing from the existing asan codebase. I understand the urgency for now, so we can move go with the current implementation, but please consider how you could reduce the number of duplications in your codebase this is creating so that asan/msan are easier to maintain long-term.

File: source/loader/layers/sanitizer/msan/msan_interceptor.cpp (hunk 17 above)
Comment (AllanZyne) [reply source/loader/layers/sanitizer/msan/msan_interceptor.cpp:166]: Sure, we noticed this problem as well, after intense period of feature developing, we will try to merge common codes to make it more easily maintain.

File: source/loader/layers/sanitizer/msan/msan_options.cpp (hunk 19 above)
Comment (AllanZyne) [reply source/loader/layers/sanitizer/msan/msan_options.cpp:25]: We consider to use a single sanitizer option in the future

**Review by pbalcer — APPROVED** [review]

**Comment by AllanZyne** [comment]
> Mostly lgtm. My biggest concern is the large amounts of copy/pasting that the new msan is doing from the existing asan codebase. I understand the urgency for now, so we can move with the current implementation, but please consider how you could reduce the number of duplications in your codebase this is creating so that asan/msan are easier to maintain long-term.

Thank you for your understanding, @pbalcer!

**Comment by kbenzie** [comment]
Once the L0 jobs complete I'll squash the commits then merge this.

**Comment by kbenzie** [comment]
After merging https://github.com/oneapi-src/unified-runtime/pull/2420 (since the CI was ready) there is a conflict on `source/loader/layers/sanitizer/asan/asan_ddi.cpp` please fix @AllanZyne and I'll merge this first thing in the morning UK time.

**Comment by AllanZyne** [comment]
> After merging #2420 (since the CI was ready) there is a conflict on `source/loader/layers/sanitizer/asan/asan_ddi.cpp` please fix @AllanZyne and I'll merge this first thing in the morning UK time.

Hi @kbenzie, I have to merge main branch to solve the conflict, which is not the commit id pointed by LLVM.
Hope the CI in LLVM works well.

**Pull request merged**